File: packages/dart/lib/src/ply_export.dart
import 'dart:io';
import 'dart:typed_data';

import 'scene.dart';

(int, int, int, int) _getMaterialRgba(Scene scene, int matIdx) {
  if (matIdx >= 0 && matIdx < scene.gltfMaterials.length) {
    final mat = scene.gltfMaterials[matIdx];
    final color = mat['baseColorFactor'];
    if (color is List && color.length >= 4) {
      final r = (color[0] * 255.0).round().clamp(0, 255);
      final g = (color[1] * 255.0).round().clamp(0, 255);
      final b = (color[2] * 255.0).round().clamp(0, 255);
      final a = (color[3] * 255.0).round().clamp(0, 255);
      return (r, g, b, a);
    }
  }
  return (200, 200, 200, 255);
}

/// Convert a baked [Scene] into ASCII PLY text format.
String toPlyAscii(Scene scene) {
  var totalVertices = 0;
  var totalFaces = 0;
  for (final prim in scene.glbPrimitives) {
    totalVertices += prim.positions.length ~/ 3;
    totalFaces += prim.indices.length ~/ 3;
  }

  final buffer = StringBuffer();
  buffer.writeln('ply');
  buffer.writeln('format ascii 1.0');
  buffer.writeln('comment Created by OpenSKP');
  buffer.writeln('element vertex $totalVertices');
  buffer.writeln('property float x');
  buffer.writeln('property float y');
  buffer.writeln('property float z');
  buffer.writeln('property float nx');
  buffer.writeln('property float ny');
  buffer.writeln('property float nz');
  buffer.writeln('property float u');
  buffer.writeln('property float v');
  buffer.writeln('property uchar red');
  buffer.writeln('property uchar green');
  buffer.writeln('property uchar blue');
  buffer.writeln('property uchar alpha');
  buffer.writeln('element face $totalFaces');
  buffer.writeln('property list uchar int vertex_indices');
  buffer.writeln('end_header');

  for (final prim in scene.glbPrimitives) {
    final rgba = _getMaterialRgba(scene, prim.materialIndex);
    final vertCount = prim.positions.length ~/ 3;
    for (var i = 0; i < vertCount; i++) {
      final px = prim.positions[i * 3].toStringAsFixed(6);
      final py = prim.positions[i * 3 + 1].toStringAsFixed(6);
      final pz = prim.positions[i * 3 + 2].toStringAsFixed(6);

      final nx = (i * 3 < prim.normals.length ? prim.normals[i * 3] : 0.0).toStringAsFixed(6);
      final ny = (i * 3 + 1 < prim.normals.length ? prim.normals[i * 3 + 1] : 0.0).toStringAsFixed(6);
      final nz = (i * 3 + 2 < prim.normals.length ? prim.normals[i * 3 + 2] : 0.0).toStringAsFixed(6);

      final u = (i * 2 < prim.uvs.length ? prim.uvs[i * 2] : 0.0).toStringAsFixed(6);
      final v = (i * 2 + 1 < prim.uvs.length ? prim.uvs[i * 2 + 1] : 0.0).toStringAsFixed(6);

      buffer.writeln('$px $py $pz $nx $ny $nz $u $v ${rgba.$1} ${rgba.$2} ${rgba.$3} ${rgba.$4}');
    }
  }

  var vertOffset = 0;
  for (final prim in scene.glbPrimitives) {
    final triCount = prim.indices.length ~/ 3;
    for (var i = 0; i < triCount; i++) {
      final i0 = prim.indices[i * 3] + vertOffset;
      final i1 = prim.indices[i * 3 + 1] + vertOffset;
      final i2 = prim.indices[i * 3 + 2] + vertOffset;
      buffer.writeln('3 $i0 $i1 $i2');
    }
    vertOffset += prim.positions.length ~/ 3;
  }

  return buffer.toString();
}

/// Convert a baked [Scene] into Little-Endian Binary PLY byte array.
Uint8List toPlyBinary(Scene scene) {
  var totalVertices = 0;
  var totalFaces = 0;
  for (final prim in scene.glbPrimitives) {
    totalVertices += prim.positions.length ~/ 3;
    totalFaces += prim.indices.length ~/ 3;
  }

  final headerText =
      'ply\n'
      'format binary_little_endian 1.0\n'
      'comment Created by OpenSKP\n'
      'element vertex $totalVertices\n'
      'property float x\n'
      'property float y\n'
      'property float z\n'
      'property float nx\n'
      'property float ny\n'
      'property float nz\n'
      'property float u\n'
      'property float v\n'
      'property uchar red\n'
      'property uchar green\n'
      'property uchar blue\n'
      'property uchar alpha\n'
      'element face $totalFaces\n'
      'property list uchar int vertex_indices\n'
      'end_header\n';

  final headerBytes = Uint8List.fromList(headerText.codeUnits);
  final vertexBytesSize = totalVertices * 36; // 8x float32 + 4x uint8
  final faceBytesSize = totalFaces * 13; // 1x uint8 + 3x int32

  final bufferSize = headerBytes.length + vertexBytesSize + faceBytesSize;
  final bytes = Uint8List(bufferSize);
  final bdata = ByteData.view(bytes.buffer);

  // Copy header
  bytes.setRange(0, headerBytes.length, headerBytes);

  var offset = headerBytes.length;

  for (final prim in scene.glbPrimitives) {
    final rgba = _getMaterialRgba(scene, prim.materialIndex);
    final vertCount = prim.positions.length ~/ 3;
    for (var i = 0; i < vertCount; i++) {
      final px = prim.positions[i * 3];
      final py = prim.positions[i * 3 + 1];
      final pz = prim.positions[i * 3 + 2];

      final nx = i * 3 < prim.normals.length ? prim.normals[i * 3] : 0.0;
      final ny = i * 3 + 1 < prim.normals.length ? prim.normals[i * 3 + 1] : 0.0;
      final nz = i * 3 + 2 < prim.normals.length ? prim.normals[i * 3 + 2] : 0.0;

      final u = i * 2 < prim.uvs.length ? prim.uvs[i * 2] : 0.0;
      final v = i * 2 + 1 < prim.uvs.length ? prim.uvs[i * 2 + 1] : 0.0;

      bdata.setFloat32(offset, px, Endian.little);
      bdata.setFloat32(offset + 4, py, Endian.little);
      bdata.setFloat32(offset + 8, pz, Endian.little);

      bdata.setFloat32(offset + 12, nx, Endian.little);
      bdata.setFloat32(offset + 16, ny, Endian.little);
      bdata.setFloat32(offset + 20, nz, Endian.little);

      bdata.setFloat32(offset + 24, u, Endian.little);
      bdata.setFloat32(offset + 28, v, Endian.little);

      bytes[offset + 32] = rgba.$1;
      bytes[offset + 33] = rgba.$2;
      bytes[offset + 34] = rgba.$3;
      bytes[offset + 35] = rgba.$4;

      offset += 36;
    }
  }

  var vertOffset = 0;
  for (final prim in scene.glbPrimitives) {
    final triCount = prim.indices.length ~/ 3;
    for (var i = 0; i < triCount; i++) {
      final i0 = prim.indices[i * 3] + vertOffset;
      final i1 = prim.indices[i * 3 + 1] + vertOffset;
      final i2 = prim.indices[i * 3 + 2] + vertOffset;

      bytes[offset] = 3;
      bdata.setInt32(offset + 1, i0, Endian.little);
      bdata.setInt32(offset + 5, i1, Endian.little);
      bdata.setInt32(offset + 9, i2, Endian.little);

      offset += 13;
    }
    vertOffset += prim.positions.length ~/ 3;
  }

  return bytes;
}

/// Export a baked [Scene] to a PLY file at [path].
void exportPly(Scene scene, String path, {bool binary = false}) {
  final file = File(path);
  final dir = file.parent;
  if (!dir.existsSync()) {
    dir.createSync(recursive: true);
  }

  if (binary) {
    final data = toPlyBinary(scene);
    file.writeAsBytesSync(data);
  } else {
    final text = toPlyAscii(scene);
    file.writeAsStringSync(text);
  }
}

File: packages/dotnet/OpenSkp/PlyExport.cs
using System;
using System.Globalization;
using System.IO;
using System.Text;

namespace OpenSkp
{
    /// <summary>
    /// PLY (Polygon File Format) exporter for baked <see cref="Scene"/> objects.
    /// Supports both ASCII and Little-Endian Binary PLY formats with positions, normals, UVs, and RGBA colors.
    /// </summary>
    public static class PlyExport
    {
        private static (byte r, byte g, byte b, byte a) GetMaterialRgba(Scene scene, int matIdx)
        {
            if (matIdx >= 0 && scene.GltfMaterials != null && matIdx < scene.GltfMaterials.Count)
            {
                var mat = scene.GltfMaterials[matIdx];
                if (mat is System.Collections.IDictionary dict && dict.Contains("baseColorFactor") && dict["baseColorFactor"] is System.Collections.IEnumerable enumerable)
                {
                    var list = new System.Collections.Generic.List<float>();
                    foreach (var item in enumerable)
                    {
                        list.Add(Convert.ToSingle(item, CultureInfo.InvariantCulture));
                    }
                    if (list.Count >= 4)
                    {
                        byte r = (byte)Math.Max(0, Math.Min(255, (int)Math.Round(list[0] * 255.0f)));
                        byte g = (byte)Math.Max(0, Math.Min(255, (int)Math.Round(list[1] * 255.0f)));
                        byte b = (byte)Math.Max(0, Math.Min(255, (int)Math.Round(list[2] * 255.0f)));
                        byte a = (byte)Math.Max(0, Math.Min(255, (int)Math.Round(list[3] * 255.0f)));
                        return (r, g, b, a);
                    }
                }
            }
            return (200, 200, 200, 255);
        }

        /// <summary>
        /// Serialize a baked <see cref="Scene"/> into ASCII PLY text format.
        /// </summary>
        /// <param name="scene">The baked scene returned by <see cref="SkpFile.BuildScene(string, SkpParseOptions?)"/>.</param>
        /// <returns>Formatted ASCII PLY text string.</returns>
        public static string ToPlyAscii(Scene scene)
        {
            if (scene == null) throw new ArgumentNullException(nameof(scene));

            int totalVertices = 0;
            int totalFaces = 0;
            foreach (var prim in scene.GlbPrimitives)
            {
                totalVertices += prim.Positions.Length / 3;
                totalFaces += prim.Indices.Length / 3;
            }

            var sb = new StringBuilder();
            sb.AppendLine("ply");
            sb.AppendLine("format ascii 1.0");
            sb.AppendLine("comment Created by OpenSKP");
            sb.AppendLine($"element vertex {totalVertices}");
            sb.AppendLine("property float x");
            sb.AppendLine("property float y");
            sb.AppendLine("property float z");
            sb.AppendLine("property float nx");
            sb.AppendLine("property float ny");
            sb.AppendLine("property float nz");
            sb.AppendLine("property float u");
            sb.AppendLine("property float v");
            sb.AppendLine("property uchar red");
            sb.AppendLine("property uchar green");
            sb.AppendLine("property uchar blue");
            sb.AppendLine("property uchar alpha");
            sb.AppendLine($"element face {totalFaces}");
            sb.AppendLine("property list uchar int vertex_indices");
            sb.AppendLine("end_header");

            foreach (var prim in scene.GlbPrimitives)
            {
                var (r, g, b, a) = GetMaterialRgba(scene, prim.MaterialIndex);
                int vertCount = prim.Positions.Length / 3;
                for (int i = 0; i < vertCount; i++)
                {
                    string px = prim.Positions[i * 3].ToString("F6", CultureInfo.InvariantCulture);
                    string py = prim.Positions[i * 3 + 1].ToString("F6", CultureInfo.InvariantCulture);
                    string pz = prim.Positions[i * 3 + 2].ToString("F6", CultureInfo.InvariantCulture);

                    float nxVal = (i * 3 < prim.Normals.Length) ? prim.Normals[i * 3] : 0.0f;
                    float nyVal = (i * 3 + 1 < prim.Normals.Length) ? prim.Normals[i * 3 + 1] : 0.0f;
                    float nzVal = (i * 3 + 2 < prim.Normals.Length) ? prim.Normals[i * 3 + 2] : 0.0f;

                    string nx = nxVal.ToString("F6", CultureInfo.InvariantCulture);
                    string ny = nyVal.ToString("F6", CultureInfo.InvariantCulture);
                    string nz = nzVal.ToString("F6", CultureInfo.InvariantCulture);

                    float uVal = (i * 2 < prim.Uvs.Length) ? prim.Uvs[i * 2] : 0.0f;
                    float vVal = (i * 2 + 1 < prim.Uvs.Length) ? prim.Uvs[i * 2 + 1] : 0.0f;

                    string u = uVal.ToString("F6", CultureInfo.InvariantCulture);
                    string v = vVal.ToString("F6", CultureInfo.InvariantCulture);

                    sb.AppendLine($"{px} {py} {pz} {nx} {ny} {nz} {u} {v} {r} {g} {b} {a}");
                }
            }

            int vertOffset = 0;
            foreach (var prim in scene.GlbPrimitives)
            {
                int triCount = prim.Indices.Length / 3;
                for (int i = 0; i < triCount; i++)
                {
                    uint i0 = prim.Indices[i * 3] + (uint)vertOffset;
                    uint i1 = prim.Indices[i * 3 + 1] + (uint)vertOffset;
                    uint i2 = prim.Indices[i * 3 + 2] + (uint)vertOffset;
                    sb.AppendLine($"3 {i0} {i1} {i2}");
                }
                vertOffset += prim.Positions.Length / 3;
            }

            return sb.ToString();
        }

        /// <summary>
        /// Serialize a baked <see cref="Scene"/> into Little-Endian Binary PLY byte array format.
        /// </summary>
        /// <param name="scene">The baked scene returned by <see cref="SkpFile.BuildScene(string, SkpParseOptions?)"/>.</param>
        /// <returns>Packed Little-Endian Binary PLY byte array.</returns>
        public static byte[] ToPlyBinary(Scene scene)
        {
            if (scene == null) throw new ArgumentNullException(nameof(scene));

            int totalVertices = 0;
            int totalFaces = 0;
            foreach (var prim in scene.GlbPrimitives)
            {
                totalVertices += prim.Positions.Length / 3;
                totalFaces += prim.Indices.Length / 3;
            }

            string headerText =
                "ply\n" +
                "format binary_little_endian 1.0\n" +
                "comment Created by OpenSKP\n" +
                $"element vertex {totalVertices}\n" +
                "property float x\n" +
                "property float y\n" +
                "property float z\n" +
                "property float nx\n" +
                "property float ny\n" +
                "property float nz\n" +
                "property float u\n" +
                "property float v\n" +
                "property uchar red\n" +
                "property uchar green\n" +
                "property uchar blue\n" +
                "property uchar alpha\n" +
                $"element face {totalFaces}\n" +
                "property list uchar int vertex_indices\n" +
                "end_header\n";

            byte[] headerBytes = Encoding.ASCII.GetBytes(headerText);
            int vertexBytesSize = totalVertices * 36;
            int faceBytesSize = totalFaces * 13;

            byte[] bytes = new byte[headerBytes.Length + vertexBytesSize + faceBytesSize];
            Array.Copy(headerBytes, 0, bytes, 0, headerBytes.Length);

            int offset = headerBytes.Length;

            foreach (var prim in scene.GlbPrimitives)
            {
                var (r, g, b, a) = GetMaterialRgba(scene, prim.MaterialIndex);
                int vertCount = prim.Positions.Length / 3;
                for (int i = 0; i < vertCount; i++)
                {
                    float px = prim.Positions[i * 3];
                    float py = prim.Positions[i * 3 + 1];
                    float pz = prim.Positions[i * 3 + 2];

                    float nx = (i * 3 < prim.Normals.Length) ? prim.Normals[i * 3] : 0.0f;
                    float ny = (i * 3 + 1 < prim.Normals.Length) ? prim.Normals[i * 3 + 1] : 0.0f;
                    float nz = (i * 3 + 2 < prim.Normals.Length) ? prim.Normals[i * 3 + 2] : 0.0f;

                    float u = (i * 2 < prim.Uvs.Length) ? prim.Uvs[i * 2] : 0.0f;
                    float v = (i * 2 + 1 < prim.Uvs.Length) ? prim.Uvs[i * 2 + 1] : 0.0f;

                    WriteFloatLE(bytes, offset, px);
                    WriteFloatLE(bytes, offset + 4, py);
                    WriteFloatLE(bytes, offset + 8, pz);

                    WriteFloatLE(bytes, offset + 12, nx);
                    WriteFloatLE(bytes, offset + 16, ny);
                    WriteFloatLE(bytes, offset + 20, nz);

                    WriteFloatLE(bytes, offset + 24, u);
                    WriteFloatLE(bytes, offset + 28, v);

                    bytes[offset + 32] = r;
                    bytes[offset + 33] = g;
                    bytes[offset + 34] = b;
                    bytes[offset + 35] = a;

                    offset += 36;
                }
            }

            int vertOffset = 0;
            foreach (var prim in scene.GlbPrimitives)
            {
                int triCount = prim.Indices.Length / 3;
                for (int i = 0; i < triCount; i++)
                {
                    int i0 = (int)prim.Indices[i * 3] + vertOffset;
                    int i1 = (int)prim.Indices[i * 3 + 1] + vertOffset;
                    int i2 = (int)prim.Indices[i * 3 + 2] + vertOffset;

                    bytes[offset] = 3;
                    WriteInt32LE(bytes, offset + 1, i0);
                    WriteInt32LE(bytes, offset + 5, i1);
                    WriteInt32LE(bytes, offset + 9, i2);

                    offset += 13;
                }
                vertOffset += prim.Positions.Length / 3;
            }

            return bytes;
        }

        private static void WriteFloatLE(byte[] buffer, int offset, float value)
        {
            byte[] fBytes = BitConverter.GetBytes(value);
            if (!BitConverter.IsLittleEndian) Array.Reverse(fBytes);
            Array.Copy(fBytes, 0, buffer, offset, 4);
        }

        private static void WriteInt32LE(byte[] buffer, int offset, int value)
        {
            byte[] iBytes = BitConverter.GetBytes(value);
            if (!BitConverter.IsLittleEndian) Array.Reverse(iBytes);
            Array.Copy(iBytes, 0, buffer, offset, 4);
        }

        /// <summary>
        /// Export a baked <see cref="Scene"/> directly to a PLY file.
        /// </summary>
        /// <param name="scene">The baked scene returned by <see cref="SkpFile.BuildScene(string, SkpParseOptions?)"/>.</param>
        /// <param name="outputPath">Destination file path (.ply).</param>
        /// <param name="binary">If true, writes binary PLY. Otherwise writes ASCII PLY.</param>
        public static void ExportPly(Scene scene, string outputPath, bool binary = false)
        {
            if (string.IsNullOrEmpty(outputPath)) throw new ArgumentException("Output path cannot be null or empty", nameof(outputPath));

            string? dir = Path.GetDirectoryName(outputPath);
            if (!string.IsNullOrEmpty(dir) && !Directory.Exists(dir))
            {
                Directory.CreateDirectory(dir);
            }

            if (binary)
            {
                byte[] data = ToPlyBinary(scene);
                File.WriteAllBytes(outputPath, data);
            }
            else
            {
                string text = ToPlyAscii(scene);
                File.WriteAllText(outputPath, text, Encoding.UTF8);
            }
        }
    }
}
